lib/Dialect/DC/Transforms/DCMaterialization.cpp

```diff
@@ -0,0 +1,142 @@
+//===- DCMaterialization.cpp - Fork/sink materialization pass ---*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the Fork/sink materialization pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/DC/DCOps.h"
+#include "circt/Dialect/DC/DCPasses.h"
+
+using namespace circt;
+using namespace dc;
+using namespace mlir;
+
+static void replaceFirstUse(Operation *op, Value oldVal, Value newVal) {
+  for (int i = 0, e = op->getNumOperands(); i < e; ++i)
+    if (op->getOperand(i) == oldVal) {
+      op->setOperand(i, newVal);
+      break;
+    }
+  return;
+}
+
+static void insertSink(Value val, OpBuilder &rewriter) {
+  rewriter.setInsertionPointAfterValue(val);
+  rewriter.create<SinkOp>(val.getLoc(), val);
+}
+
+static void insertFork(Value result, OpBuilder &rewriter) {
+  // Get successor operations
+  std::vector<Operation *> opsToProcess;
+  for (auto &u : result.getUses())
+    opsToProcess.push_back(u.getOwner());
+
+  // Insert fork after op
+  rewriter.setInsertionPointAfterValue(result);
+  auto forkSize = opsToProcess.size();
+  auto newFork = rewriter.create<ForkOp>(result.getLoc(), result, forkSize);
+
+  // Modify operands of successor
+  // opsToProcess may have multiple instances of same operand
+  // Replace uses one by one to assign different fork outputs to them
+  for (auto [op, forkRes] : llvm::zip(opsToProcess, newFork->getResults()))
+    replaceFirstUse(op, result, forkRes);
+}
+
+// Insert Fork Operation for every operation and function argument with more
+// than one successor.
+static LogicalResult addForkOps(Region &r, OpBuilder &rewriter) {
+  for (Operation &op : r.getOps()) {
+    // Ignore terminators, and don't add Forks to Forks.
+    if (op.getNumSuccessors() == 0 && !isa<ForkOp>(op)) {
+      for (auto result : op.getResults()) {
+        // If there is a result, it is used more than once, and it is a DC type,
+        // fork it!
+        if (!result.use_empty() && !result.hasOneUse() &&
+            result.getType().isa<dc::TokenType, dc::ValueType>())
+          insertFork(result, rewriter);
+      }
+    }
+  }
+
+  for (auto barg : r.front().getArguments())
+    if (!barg.use_empty() && !barg.hasOneUse())
+      insertFork(barg, rewriter);
+
+  return success();
+}
+
+namespace circt {
+namespace dc {
+
+// Create sink for every unused result
+LogicalResult addSinkOps(Region &r, OpBuilder &rewriter) {
+  for (Block &block : r) {
```

Comment:
Why do we iterate over all blocks here but not in `addForkOps`? I would expect the logic to be the same.

Reply:
Shouldn't `walk` be used to get to all the blocks in a design recursively?